In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
#importing First Datset
Survive=pd.read_csv("titanic.csv")
Survive.sample(5)

PassengerId  Survived  Pclass  \
786          787         1       3   
139          140         0       1   
589          590         0       3   
566          567         0       3   
8              9         1       3   

                                                  Name     Sex   Age  SibSp  \
786                          Sjoblom, Miss. Anna Sofia  female  18.0      0   
139                                 Giglio, Mr. Victor    male  24.0      0   
589                                Murdlin, Mr. Joseph    male   NaN      0   
566                               Stoytcheff, Mr. Ilia    male  19.0      0   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   

     Parch      Ticket     Fare Cabin Embarked  
786      0     3101265   7.4958   NaN        S  
139      0    PC 17593  79.2000   B86        C  
589      0  A./5. 3235   8.0500   NaN        S  
566      0      349205   7.8958   NaN        S  
8        2      347742  11.1333   NaN        S

In [3]:
Survive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [4]:
#Removing The Unnessary Columns Not Required for Our Analysis
Survive_New=Survive.drop(['PassengerId','Name','SibSp','Parch','Embarked','Cabin','Ticket'], axis='columns')
Survive_New

Survived  Pclass     Sex   Age     Fare
0           0       3    male  22.0   7.2500
1           1       1  female  38.0  71.2833
2           1       3  female  26.0   7.9250
3           1       1  female  35.0  53.1000
4           0       3    male  35.0   8.0500
..        ...     ...     ...   ...      ...
886         0       2    male  27.0  13.0000
887         1       1  female  19.0  30.0000
888         0       3  female   NaN  23.4500
889         1       1    male  26.0  30.0000
890         0       3    male  32.0   7.7500

[891 rows x 5 columns]

In [5]:
#Making Dummy Values(Removing categorical Values)
New_survive=pd.get_dummies(Survive_New.Sex)
New_survive=New_survive.astype(int)

In [6]:
#Making The Final Dataset
Final_Survive=pd.concat([Survive_New,New_survive],axis='columns')
Final_Survive=Final_Survive.drop('Sex',axis="columns")
Final_Survive

Survived  Pclass   Age     Fare  female  male
0           0       3  22.0   7.2500       0     1
1           1       1  38.0  71.2833       1     0
2           1       3  26.0   7.9250       1     0
3           1       1  35.0  53.1000       1     0
4           0       3  35.0   8.0500       0     1
..        ...     ...   ...      ...     ...   ...
886         0       2  27.0  13.0000       0     1
887         1       1  19.0  30.0000       1     0
888         0       3   NaN  23.4500       1     0
889         1       1  26.0  30.0000       0     1
890         0       3  32.0   7.7500       0     1

[891 rows x 6 columns]

In [7]:
#Finding The Unnessary Values
Final_Survive.isnull().sum()

Survived      0
Pclass        0
Age         177
Fare          0
female        0
male          0
dtype: int64

In [8]:
# We See That Age Column has Unnessary values So Let's Fill Them With The Mean of Age Column
Avearge_Age=Final_Survive.Age.mean()
Final_Survive=Final_Survive.fillna(int(Avearge_Age))
Final_Survive

Survived  Pclass   Age     Fare  female  male
0           0       3  22.0   7.2500       0     1
1           1       1  38.0  71.2833       1     0
2           1       3  26.0   7.9250       1     0
3           1       1  35.0  53.1000       1     0
4           0       3  35.0   8.0500       0     1
..        ...     ...   ...      ...     ...   ...
886         0       2  27.0  13.0000       0     1
887         1       1  19.0  30.0000       1     0
888         0       3  29.0  23.4500       1     0
889         1       1  26.0  30.0000       0     1
890         0       3  32.0   7.7500       0     1

[891 rows x 6 columns]

In [9]:
#Checking Again
Final_Survive.isnull().sum()

Survived    0
Pclass      0
Age         0
Fare        0
female      0
male        0
dtype: int64

In [10]:
# Getting The Variables Assigned
inputs=Final_Survive[['Pclass','Age','Fare','female']]
output=Final_Survive.Survived

In [11]:
#Importing Test_Train_Split For 75% Training sample
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(inputs,output,train_size=0.75,random_state=42) #For 75% Training sample

In [12]:
#Importing Test_Train_Split For 66% Training sample
from sklearn.model_selection import train_test_split
x1_train,x1_test,y1_train,y1_test=train_test_split(inputs,output,train_size=0.66,random_state=42) #For 66% Training sample

In [13]:
# Making K Nearest Neighbours for 75% Training Set
#Importing K Nearsest Neighbors Algorithm
from sklearn.neighbors import KNeighborsClassifier
K_predict75=KNeighborsClassifier(n_neighbors=3)
K_predict75.fit(x_train,y_train)

KNeighborsClassifier(n_neighbors=3)

In [14]:
#Predicting Values
K_predict75.predict(x_test)

array([0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0,
       0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0,
       0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0,
       0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 1,
       1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 1, 0,
       0, 1, 0], dtype=int64)

In [15]:
#Predicting Scores
K_predict75.score(x_train,y_train)*100
#Its 83.52 Percent Accurate

83.53293413173652

In [16]:
# Making K Nearest Neighbours for 66% Training Set
#Importing K Nearsest Neighbors Algorithm
from sklearn.neighbors import KNeighborsClassifier
K_predict66=KNeighborsClassifier(n_neighbors=3)
K_predict66.fit(x1_train,y1_train)

KNeighborsClassifier(n_neighbors=3)

In [17]:
#Predicting Values
K_predict66.predict(x1_test)

array([0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0,
       0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0,
       0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 1,
       1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0,
       0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1,
       1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0,

In [18]:
#Predicting Scores
K_predict66.score(x1_train,y1_train)*100
#Its 81.63 Percent Accurate

81.63265306122449

#### In K Nearest Neighbours 75% Training Set Performs Better Than 66% Training Set

In [19]:
# Making Decision Tree Classifier for 75% Training Set
from sklearn.tree import DecisionTreeClassifier
DT_Predict75=DecisionTreeClassifier()
DT_Predict75.fit(x_train,y_train)

DecisionTreeClassifier()

In [20]:
#Predicting Values
DT_Predict75.predict(x_test)

array([0, 1, 1, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1,
       0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0,
       1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0,
       0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0,
       0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 0,
       1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0,
       0, 1, 0], dtype=int64)

In [21]:
#Predicting Scores
DT_Predict75.score(x_train,y_train)*100
#It's 97.45% Accurate

97.45508982035929

In [22]:
# Making Decision Tree Classifier for 66% Training Set
from sklearn.tree import DecisionTreeClassifier
DT_Predict66=DecisionTreeClassifier()
DT_Predict66.fit(x1_train,y1_train)

DecisionTreeClassifier()

In [23]:
#Predicting Values
DT_Predict66.predict(x_test)

array([0, 0, 1, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1,
       0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0,
       1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0,
       0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0,
       0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 0,
       0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0,
       1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0,
       0, 1, 0], dtype=int64)

In [24]:
#Predicting Scores
DT_Predict66.score(x1_train,y1_train)*100
#It's 97.27% Accurate

97.27891156462584

#### In Decision tree Classifier 75% Training Set Performs Ever So Slighty (0.25% More) Better Than 66% Training Set

In [25]:
# Making Naive Bayes Classifier for 75% Training Set
from sklearn.naive_bayes import GaussianNB
NB_Model75= GaussianNB()
NB_Model75.fit(x_train,y_train)

GaussianNB()

In [26]:
#Predicting Values
NB_Model75.predict(x_test)

array([0, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1,
       0, 0, 0, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0,
       1, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 0,
       0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1,
       0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0,
       1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 0,
       1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0,
       0, 1, 0], dtype=int64)

In [27]:
#Predicting Scores
NB_Model75.score(x_train,y_train)*100
#It's 78.14% Accurate

78.1437125748503

In [28]:
# Making Naive Bayes Classifier for 66% Training Set
from sklearn.naive_bayes import GaussianNB
NB_Model66= GaussianNB()
NB_Model66.fit(x1_train,y1_train)

GaussianNB()

In [29]:
#Predicting Values
NB_Model66.predict(x1_test)

array([0, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1,
       0, 0, 0, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0,
       1, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 0,
       0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1,
       0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0,
       1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 0,
       1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0,
       0, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 1,
       0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0,
       0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1,

In [30]:
#Predicting Scores
NB_Model66.score(x1_train,y1_train)*100
#It's 77.21% Accurate

77.21088435374149

#### In Naive Bayes Classifier 75% Training Set Performs Slighty Better(Almost 1%) Than 66% Training Set

### Importing the Second Dataset

In [31]:
Gender_Classification=pd.read_csv("gender_classification.csv")
Gender_Classification.sample(5)

long_hair  forehead_width_cm  forehead_height_cm  nose_wide  nose_long  \
214           1               13.1                 5.7          1          1   
424           1               13.2                 6.6          0          1   
2344          1               13.1                 6.5          1          1   
1043          1               11.7                 6.3          0          0   
4469          1               14.9                 5.9          1          1   

      lips_thin  distance_nose_to_lip_long  gender  
214           1                          1    Male  
424           1                          1    Male  
2344          1                          1    Male  
1043          0                          0  Female  
4469          1                          1    Male

In [32]:
#Checking Size
Gender_Classification.shape

(5001, 8)

In [33]:
Gender_Classification.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5001 entries, 0 to 5000
Data columns (total 8 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   long_hair                  5001 non-null   int64  
 1   forehead_width_cm          5001 non-null   float64
 2   forehead_height_cm         5001 non-null   float64
 3   nose_wide                  5001 non-null   int64  
 4   nose_long                  5001 non-null   int64  
 5   lips_thin                  5001 non-null   int64  
 6   distance_nose_to_lip_long  5001 non-null   int64  
 7   gender                     5001 non-null   object 
dtypes: float64(2), int64(5), object(1)
memory usage: 312.7+ KB


In [34]:
#Checking Null Values
Gender_Classification.isnull().sum()

long_hair                    0
forehead_width_cm            0
forehead_height_cm           0
nose_wide                    0
nose_long                    0
lips_thin                    0
distance_nose_to_lip_long    0
gender                       0
dtype: int64

In [35]:
#Making Intputs and Outputs
Predicting_From=Gender_Classification.iloc[:,0:7]
Predicted=Gender_Classification.gender

In [36]:
#Importing Test_Train_Split For 75% Training sample For Gender_Classification Dataset
from sklearn.model_selection import train_test_split
Gender_x_train,Gender_x_test,Gender_y_train,Gender_y_test=train_test_split(Predicting_From,Predicted ,train_size=0.75,random_state=42) 
#For 75% Training sample

In [37]:
#Importing Test_Train_Split For 66% Training sample For Gender_Classification Dataset
from sklearn.model_selection import train_test_split
Gender_x1_train,Gender_x1_test,Gender_y1_train,Gender_y1_test=train_test_split(Predicting_From,Predicted,train_size=0.66,random_state=42) 
#For 66% Training sample

In [38]:
# Making K Nearest Neighbours for 75% Training Set For Gender_Classification Dataset
#Importing K Nearsest Neighbors Algorithm
from sklearn.neighbors import KNeighborsClassifier
K_New_predict75=KNeighborsClassifier(n_neighbors=3)
K_New_predict75.fit(Gender_x_train,Gender_y_train)

KNeighborsClassifier(n_neighbors=3)

In [39]:
#Predicting Values
K_New_predict75.predict(Gender_x_test)

array(['Male', 'Female', 'Female', ..., 'Male', 'Male', 'Male'],
      dtype=object)

In [40]:
#Predicting Scores
K_New_predict75.score(Gender_x_train,Gender_y_train)*100
#Its 98.26% Accurate

98.26666666666667

In [41]:
# Making K Nearest Neighbours for 66% Training Set For Gender_Classification Dataset
#Importing K Nearsest Neighbors Algorithm
from sklearn.neighbors import KNeighborsClassifier
K_New_predict66=KNeighborsClassifier(n_neighbors=3)
K_New_predict66.fit(Gender_x1_train,Gender_y1_train)

KNeighborsClassifier(n_neighbors=3)

In [42]:
#Predicting Values
K_New_predict66.predict(Gender_x1_test)

array(['Male', 'Female', 'Female', ..., 'Female', 'Female', 'Female'],
      dtype=object)

In [43]:
#Predicting Scores
K_New_predict66.score(Gender_x1_train,Gender_y1_train)*100
#Its 98.27% Accurate

98.27272727272728

### In K Nearest Neighbours 66% Training Set Performs Very Very Slightly(0.01%) Better Than 75% Training Set

In [44]:
# Making Decision Tree Classifier for 75% Training Set For Gender_Classification Dataset
from sklearn.tree import DecisionTreeClassifier
New_DT_Predict75=DecisionTreeClassifier()
New_DT_Predict75.fit(Gender_x_train,Gender_y_train)

DecisionTreeClassifier()

In [45]:
#Predicting Values
New_DT_Predict75.predict(Gender_x_test)

array(['Male', 'Female', 'Female', ..., 'Male', 'Male', 'Male'],
      dtype=object)

In [46]:
#Predicting Scores
New_DT_Predict75.score(Gender_x_train,Gender_y_train)*100
#It's 99.83% Accurate

99.83999999999999

In [47]:
# Making Decision Tree Classifier for 66% Training Set For Gender_Classification Dataset
from sklearn.tree import DecisionTreeClassifier
New_DT_Predict66=DecisionTreeClassifier()
New_DT_Predict66.fit(Gender_x1_train,Gender_y1_train)

DecisionTreeClassifier()

In [48]:
#Predicting Values
New_DT_Predict66.predict(Gender_x1_test)

array(['Male', 'Female', 'Female', ..., 'Female', 'Male', 'Female'],
      dtype=object)

In [49]:
#Predicting Scores
New_DT_Predict66.score(Gender_x1_train,Gender_y1_train)*100
#It's 99.90% Accurate

99.90909090909092

### In Decision Tree Classifier 66% Training Set Performs(0.6)% Better Than 75% Training Set

In [50]:
# Making Naive Bayes Classifier for 75% Training Set For Gender_Classification Dataset
from sklearn.naive_bayes import GaussianNB
New_NB_Model75= GaussianNB()
New_NB_Model75.fit(Gender_x_train,Gender_y_train)

GaussianNB()

In [51]:
#Predicting Values
New_NB_Model75.predict(Gender_x_test)

array(['Male', 'Female', 'Female', ..., 'Male', 'Male', 'Male'],
      dtype='<U6')

In [52]:
#Predicting Scores
New_NB_Model75.score(Gender_x_train,Gender_y_train)*100
#It's 97.04% Accurate

97.04

In [53]:
# Making Naive Bayes Classifier for 66% Training Set For Gender_Classification Dataset
from sklearn.naive_bayes import GaussianNB
New_NB_Model66= GaussianNB()
New_NB_Model66.fit(Gender_x1_train,Gender_y1_train)

GaussianNB()

In [54]:
#Predicting Values
New_NB_Model66.predict(Gender_x1_test)

array(['Male', 'Female', 'Female', ..., 'Female', 'Female', 'Female'],
      dtype='<U6')

In [55]:
#Predicting Scores
New_NB_Model66.score(Gender_x1_train,Gender_y1_train)*100
#It's 97.15% Accurate

97.15151515151516

### In Navie Bayes Classifier 66% Training Set Performs(0.1)% Better Than 75% Training Set

### Using Cross Validation Method

In [56]:
from sklearn.model_selection import KFold 
from sklearn.model_selection import cross_val_score

In [66]:
cv = KFold(n_splits=7, random_state=1, shuffle=True)

In [78]:
scores = cross_val_score(K_New_predict75, Predicting_From,Predicted, scoring='accuracy', cv=cv, n_jobs=-1)
# report performance
print('Accuracy Of The KNearest Neighbours is : %.3f' % (np.mean(scores)*100))

Accuracy Of The KNearest Neighbours  is : 96.581


#### We Observe That Using Holdout Method We Got a Better Result For KNearset Neighbours (98.26%) Which Is Better Than Cross Validation (96.58%)

In [76]:
scores = cross_val_score(New_DT_Predict75, Predicting_From,Predicted, scoring='accuracy', cv=cv, n_jobs=-1)
# report performance
print('Accuracy Of The Decision Tree Classifier is  %.3f' % (np.mean(scores)*100))

Accuracy: 96.301


#### We Observe That Using Holdout Method We Got a Better Result For Decision tree Classifier (99.83%) Which Is Better Than Cross Validation (96.30%)

In [79]:
scores = cross_val_score(New_NB_Model75, Predicting_From,Predicted, scoring='accuracy', cv=cv, n_jobs=-1)
# report performance
print('Accuracy Of The Naive Bayes Classifier is  %.3f' % (np.mean(scores)*100))

Accuracy Of The Naive Bayes Classifier is  97.141


#### We Observe That Using Holdout Method We Got Almost Same/Worse Result For Naive Bayes Classifier (97.04%) Which Is Worse Than Cross Validation (97.141%)

## Using random Subsampling

In [105]:
# Making a Function to Splitting the dataset into Training and Testing Data

def split_data(test_size, random_state):
    from sklearn.model_selection import train_test_split
    X2_train, X2_test, y2_train, y2_test = train_test_split(inputs,output,test_size=test_size, random_state=random_state, shuffle=True)
    # Standardizing the Dataset
    from sklearn.preprocessing import StandardScaler
    sc = StandardScaler()# Splitting the dataset into Training and Testing Data
    X_train = sc.fit_transform(X2_train)
    X_test = sc.transform(X2_test)
    
    gnb = GaussianNB()
    y_pred = gnb.fit(X2_train, y2_train)

    y_pred_gnb = gnb.predict(X2_test)
    print("\nGaussian Naive Bayes model accuracy(in %):", accuracy_score(y2_test, y_pred_gnb)*100)
    accuracies_gnb.append(accuracy_score(y2_test, y_pred_gnb))
    
    # K-Nearest Neighbor Classifier

    knn = KNeighborsClassifier(n_neighbors=7)
    knn.fit(X2_train, y2_train)

    y_pred_knn = knn.predict(X2_test)
    print("K-Nearest Neighbors model accuracy(in %):", accuracy_score(y2_test, y_pred_knn)*100)
    accuracies_knn.append(accuracy_score(y2_test, y_pred_knn))
    
    # Decision Tree Classifier

    dtree = DecisionTreeClassifier()
    dtree.fit(X2_train, y2_train)

    y_pred_dtree = dtree.predict(X2_test)
    print("Decision Tree model accuracy(in %):", accuracy_score(y2_test, y_pred_dtree)*100)
    accuracies_dtree.append(accuracy_score(y2_test, y_pred_dtree))
    
    

In [106]:
accuracies_gnb = []
accuracies_dtree = []
accuracies_knn = []

In [107]:
from sklearn.metrics import accuracy_score
num = 42
for i in range(10):
    print(f"\nIter {i}:")
    split_data(test_size = 0.75, random_state = num)
    num += 10
    print("-"*100)   


Iter 0:

Gaussian Naive Bayes model accuracy(in %): 77.87742899850522
K-Nearest Neighbors model accuracy(in %): 67.2645739910314
Decision Tree model accuracy(in %): 73.69207772795217
----------------------------------------------------------------------------------------------------

Iter 1:

Gaussian Naive Bayes model accuracy(in %): 77.87742899850522
K-Nearest Neighbors model accuracy(in %): 69.80568011958147
Decision Tree model accuracy(in %): 74.88789237668162
----------------------------------------------------------------------------------------------------

Iter 2:

Gaussian Naive Bayes model accuracy(in %): 79.37219730941703
K-Nearest Neighbors model accuracy(in %): 66.51718983557548
Decision Tree model accuracy(in %): 75.63527653213752
----------------------------------------------------------------------------------------------------

Iter 3:

Gaussian Naive Bayes model accuracy(in %): 77.27952167414051
K-Nearest Neighbors model accuracy(in %): 66.06875934230195
Decision Tre

In [108]:
print("Final Accuracy of Naive Bayes of 75% Model: ", np.mean(accuracies_gnb)*100)
print("Final Accuracy of KNN of 75% Model: ", np.mean(accuracies_knn)*100)
print("Final Accuracy of Decision Tree of 75% Model: ", np.mean(accuracies_dtree)*100)

Final Accuracy of Naive Bayes:  76.80119581464872
Final Accuracy of KNN:  66.9506726457399
Final Accuracy of Decision Tree:  75.09715994020927


In [109]:
from sklearn.metrics import accuracy_score
num = 42
for i in range(10):
    print(f"\nIter {i}:")
    split_data(test_size = 0.66, random_state = num)
    num += 10
    print("-"*100)   


Iter 0:

Gaussian Naive Bayes model accuracy(in %): 79.11714770797963
K-Nearest Neighbors model accuracy(in %): 68.59083191850594
Decision Tree model accuracy(in %): 76.5704584040747
----------------------------------------------------------------------------------------------------

Iter 1:

Gaussian Naive Bayes model accuracy(in %): 76.06112054329371
K-Nearest Neighbors model accuracy(in %): 68.76061120543294
Decision Tree model accuracy(in %): 71.47707979626486
----------------------------------------------------------------------------------------------------

Iter 2:

Gaussian Naive Bayes model accuracy(in %): 78.26825127334465
K-Nearest Neighbors model accuracy(in %): 67.06281833616299
Decision Tree model accuracy(in %): 77.58913412563668
----------------------------------------------------------------------------------------------------

Iter 3:

Gaussian Naive Bayes model accuracy(in %): 76.7402376910017
K-Nearest Neighbors model accuracy(in %): 68.76061120543294
Decision Tree

In [111]:
print("Final Accuracy of Naive Bayes of 66% Model: ", np.mean(accuracies_gnb)*100)
print("Final Accuracy of KNN of 66% Model: ", np.mean(accuracies_knn)*100) 
print("Final Accuracy of Decision Tree of 66% Model: ", np.mean(accuracies_dtree)*100)

Final Accuracy of Naive Bayes of 66% Model:  76.61791539459092
Final Accuracy of KNN of 66% Model:  67.49910542303972
Final Accuracy of Decision Tree of 66% Model:  75.1037582383559


### In This Random Sampling Method Both 75% Training Data and 66% Training Data Performs Almost The Same With 66% Having a 0.1% Edge In almost All Models

## We Now Observe that Holdout method Got The Best Results out of All Three Training and Testing Splitting Techiniques Used